### PR TITLE
Support running tests with alternative JDBC drivers

### PR DESCRIPTION
**Pull request by brian-rickman**
With these changes, run_on_docker.sh can be used to run product-tests with the open source JDBC driver (default) or with the Teradata/Simba JDBC driver.  To use a different JDBC driver, edit locations.sh.

I have verified that these changes do not cause any test failures when the open source JDBC driver is used.  

Several tests fail when run with the Simba JDBC driver due to differences in the way the two JDBC drivers report column datatypes.  I will fix that in a later PR.

We will need to modify the configuration of our Kerberos Docker image before it can be used with the Simba JDBC driver.  That also will be done in a separate PR.

@ebd2 @anusudarsan @arhimondr @sanjay990

### CODE REVIEW

**Comment by brian-rickman** [comment]
Testing a fix for the Travis failure now...

File: presto-product-tests/bin/locations.sh
Comment (ebd2) [inline presto-product-tests/bin/locations.sh]: Clarification on why it works better, please.

Comment (ArturGajowy) [reply presto-product-tests/bin/locations.sh]: It's not necessarily a jar, it's a classpath entry (like the PRODUCT_TEST_ROOT you use here).
Also, it does not say which jdbc driver it is. Please rename the variable PRESTO_JDBC_DRIVER_CLASSPATH

EDIT: actually, it should always be a jar, because presto-tests does not have the product tests's `target` dir, just the jar. So let's name it PRESTO_JDBC_DRIVER_JAR and default to $PRODUCT_TEST_JAR

EDIT2: that's another reason to put this variable in compose-commons, as that's the place PRODUCT_TEST_JAR has its default defined.

Comment (ArturGajowy) [reply presto-product-tests/bin/locations.sh]: Also, this should be overridable using an environment variable (without code change). Please use the default value construct, as can be seen in compose-commons.sh

File: presto-product-tests/bin/locations.sh
Comment (ArturGajowy) [inline presto-product-tests/bin/locations.sh]: Let's just have one TEMPTO_CONFIG_YAML variable, since tempto can only be run with one config. Let's define the variable value in each profile's `compose.sh` file, so that it's defined when using the compose.sh scripts standalone. Let's make it overridable using env vars.

File: presto-product-tests/bin/locations.sh
Comment (ArturGajowy) [inline presto-product-tests/bin/locations.sh]: Please move this to compose-commons.sh. The value of this variable is of no interest to other places that `source locations.sh`. It's only needed for filling the variables in compose files, so let's have it in `compose-commons`.

File: presto-product-tests/bin/locations.sh
Comment (ArturGajowy) [inline presto-product-tests/bin/locations.sh]: Please remove the commented-out code, along with the comment. Please document the PRESTO_JDBC_DRIVER_JAR env variable in README.md, near the info on PRODUCT_TESTS_JAR et al.

File: presto-product-tests/conf/docker/common/kerberos.yml
Comment (ArturGajowy) [inline presto-product-tests/conf/docker/common/kerberos.yml]: This line can be removed, as the entry in standard.yml will be parametrized with TEMPTO_CONFIG_YAML.

File: presto-product-tests/conf/tempto/tempto-configuration-for-docker-default.yaml
Comment (ArturGajowy) [inline presto-product-tests/conf/tempto/tempto-configuration-for-docker-default.yaml]: Why include ldap configuration by default? I'd say let's keep the separate tempto configs in each profile's dir.

Also, this change should be a separate commit. But let's stick to tempto-config-per-environment. If we find the config duplication a problem, we can think of merging the configs in a way similar to what docker-copose does with multiple compose files.

File: presto-product-tests/conf/tempto/tempto-configuration-for-docker-kerberos-simba.yaml
Comment (ArturGajowy) [inline presto-product-tests/conf/tempto/tempto-configuration-for-docker-kerberos-simba.yaml]: that's hell of a features in one config! :) Does it work without kerberos? (with the non-kerberized docker image?) If not, I insist we have that variant of the config too.

**Review by ArturGajowy — CHANGES_REQUESTED** [review]

**Comment by ArturGajowy** [comment]
This should also be tested with presto-checks, since it makes significant changes to how product tests are run. Ping me offline and I'll help you with triggering it and maybe setting up a branch in presto-checks that runs checks with the simba driver :)

File: presto-product-tests/conf/tempto/tempto-configuration-for-docker-kerberos-simba.yaml
Comment (brian-rickman) [reply presto-product-tests/conf/tempto/tempto-configuration-for-docker-kerberos-simba.yaml]: This file is a copy of tempto-configuration-for-docker-kerberos.yaml.  The only parts that are different are the JDBC class and JDBC URL.  If there is a way to override those two lines without duplicating the file, I would much prefer to do that.

File: presto-product-tests/conf/tempto/tempto-configuration-for-docker-default.yaml
Comment (brian-rickman) [reply presto-product-tests/conf/tempto/tempto-configuration-for-docker-default.yaml]: Since the ldap config variables are only used by the ldap tests, it doesn't do any harm to have them in the standard config, and it gives us one less (mostly identical) config file to maintain.

If we could merge tempto config files at runtime, I would much prefer to do that.

File: presto-product-tests/conf/docker/common/kerberos.yml
Comment (brian-rickman) [reply presto-product-tests/conf/docker/common/kerberos.yml]: I tried that at first, but could not find a way to duplicate the original default behavior, where using run_on_docker.sh with a kerberos profile would run the tests using the open source JDBC driver without the user having to make any additional config or environment changes. 

If maintaining that behavior is not a requirement, then I agree a single environment variable in standard.yml will be better.

File: presto-product-tests/bin/locations.sh
Comment (brian-rickman) [reply presto-product-tests/bin/locations.sh]: ok.  will do.

File: presto-product-tests/bin/locations.sh
Comment (brian-rickman) [reply presto-product-tests/bin/locations.sh]: I'll work on this.  Might ping you off-line for more details on the proper way to implement it.  I appreciate all the feedback.

File: presto-product-tests/conf/tempto/tempto-configuration-for-docker-kerberos-simba.yaml
Comment (ArturGajowy) [reply presto-product-tests/conf/tempto/tempto-configuration-for-docker-kerberos-simba.yaml]: Did you check if it'll work on the non-kerberized image (cdh5-hive)?

File: presto-product-tests/conf/tempto/tempto-configuration-for-docker-default.yaml
Comment (ArturGajowy) [reply presto-product-tests/conf/tempto/tempto-configuration-for-docker-default.yaml]: Here's what I found on the topic: http://stackoverflow.com/a/25632724/52142. The only issue would be having perl with that package installed on each dev's machine. That could be done by publishing a busybox/alpine docker image with perl and that module preinstalled, and maybe a wrapper script. So usage would be sth like:

`docker run teradatalabs/yaml-merge file1.yml file2.yml > merged.yml`

Alternatively we could implement yaml merging in tempto and allow passing multiple config files.

Comment (ArturGajowy) [reply presto-product-tests/conf/tempto/tempto-configuration-for-docker-default.yaml]: That all seems like a scope for a separate PR though. So, please either revert the config merge, or separate it to its own commit.

File: presto-product-tests/conf/tempto/tempto-configuration-for-docker-kerberos-simba.yaml
Comment (brian-rickman) [reply presto-product-tests/conf/tempto/tempto-configuration-for-docker-kerberos-simba.yaml]: That config file would not, but it is only used with the kerberos images.

Comment (brian-rickman) [reply presto-product-tests/conf/tempto/tempto-configuration-for-docker-kerberos-simba.yaml]: This scenario makes a stronger case for the need to merge Tempto config files at runtime.

File: presto-product-tests/conf/docker/common/kerberos.yml
Comment (ArturGajowy) [reply presto-product-tests/conf/docker/common/kerberos.yml]: I think the jdbc driver setting is orthogonal to the profile used. We should allow driver overrides regardless of the profile.

Comment (brian-rickman) [reply presto-product-tests/conf/docker/common/kerberos.yml]: Yes, that was my intention.  I was also trying to ensure that the default worked with all profiles.  I think the default will not work for the kerberos profiles if we make the change you are suggesting.

File: presto-product-tests/conf/tempto/tempto-configuration-for-docker-kerberos-simba.yaml
Comment (ArturGajowy) [reply presto-product-tests/conf/tempto/tempto-configuration-for-docker-kerberos-simba.yaml]: I think should have another config file for running simba driver tests without kerberos

File: presto-product-tests/conf/tempto/tempto-configuration-for-docker-default.yaml
Comment (brian-rickman) [reply presto-product-tests/conf/tempto/tempto-configuration-for-docker-default.yaml]: Having thought about it more, I agree that this change is not directly related to this PR, so I'll revert it.

Comment (brian-rickman) [reply presto-product-tests/conf/tempto/tempto-configuration-for-docker-default.yaml]: Tempto already has yaml merge logic, since the "local" config file overrides some of the settings in the global yaml file.  So the best solution is probably to have tempto accept multiple values for "--config-local".  The only negative there is we are splitting this project across three git repos (presto, tempto, and docker).

File: presto-product-tests/conf/docker/multinode/compose.sh
Comment (ArturGajowy) [inline presto-product-tests/conf/docker/multinode/compose.sh]: We should use `${PRODUCT_TESTS_ROOT}` instead of `.` in the default value (hence the build failure https://travis-ci.org/Teradata/presto/jobs/163555938)

Comment (ArturGajowy) [reply presto-product-tests/conf/docker/multinode/compose.sh]: Also, please try some line breaks to make it read better (e.g. after `$(` ? probably some escaping is needed)

File: presto-product-tests/conf/docker/common/compose-commons.sh
Comment (ArturGajowy) [inline presto-product-tests/conf/docker/common/compose-commons.sh]: PRESTO_VERSION is undefined until we source presto.env. Please do it with an `if` statement as above - else it will blow up in presto-checks.

Comment (ArturGajowy) [reply presto-product-tests/conf/docker/common/compose-commons.sh]: Alternatively we could create a function `get_presto_version` that would source presto.env and echo PRESTO_VERSION. Then we could use the function call as a default vaule (that is, as long as ${VARIABLE:-default_value} syntax evaluates the defaults lazily). Then all usages of `source presto.env` should be refactored

File: presto-product-tests/conf/docker/multinode/compose.sh
Comment (brian-rickman) [reply presto-product-tests/conf/docker/multinode/compose.sh]: got it

File: presto-product-tests/conf/docker/common/compose-commons.sh
Comment (brian-rickman) [reply presto-product-tests/conf/docker/common/compose-commons.sh]: ok

File: presto-product-tests/conf/docker/common/compose-commons.sh
Comment (ArturGajowy) [inline presto-product-tests/conf/docker/common/compose-commons.sh]: Please add a newline for consistency

File: presto-product-tests/README.md
Comment (ArturGajowy) [inline presto-product-tests/README.md]: Please mention the tempto config yaml here

**Comment by ArturGajowy** [comment]
I've seen the tests pass. Please apply the last two comments and squash the commits. Please also reformat the commit message not to extend past a hard limit of 70 chars per row, which is the rule in upstream fork (otherwise github UI does some wrapping).

**Comment by ArturGajowy** [comment]
I'll spin up a presto-checks build now.

**Comment by ArturGajowy** [comment]
I've run the tests using simba driver, with some errors:
https://travis-ci.org/Teradata/presto-checks/builds/163983543

Some might need simba interaction, but need to be triaged anyway before we merge this.

We're also lacking a simba + kerberos tempto config, please add one in the path ./presto-product-tests/conf/tempto/tempto-configuration-for-docker-kerberos-simba.yaml

**Comment by ArturGajowy** [comment]
Also, please remember to add the basic (no kerberos) simba config when issuing the upstream PR.

**Comment by ArturGajowy** [comment]
presto-checks build for standard jdbc driver after merging this change is here: https://travis-ci.org/Teradata/presto-checks/builds/163997062

**Comment by brian-rickman** [comment]
Is presto-checks just running the standard product-tests?

The tests that fail with the Simba JDBC driver appear to be due to known differences between the Simba and Open Source JDBC drivers.  I will fix those tests, but I think that should be a separate PR.

simba + kerberos will require new a new Docker image for Kerberos.  It's going to take some time to get it to work.  That will require new PRs for product-tests and docker images.

**Comment by ArturGajowy** [comment]
> Is presto-checks just running the standard product-tests?

Standard product tests + storage formats + big_query (in .2 job) + auth_mechanisms (incl kerberos - in job .3)

**Comment by ArturGajowy** [comment]
> The tests that fail with the Simba JDBC driver appear to be due to known differences between the Simba and Open Source JDBC drivers. I will fix those tests, but I think that should be a separate PR.

OK. I think this PR is feature-complete. We can file an FB pr even with those tests failing - it's the customizability that's the scope of this PR.

**Comment by ArturGajowy** [comment]
> simba + kerberos will require new a new Docker image for Kerberos. It's going to take some time to get it to work. That will require new PRs for product-tests and docker images.

won't the cdh5-hive-kerberized image do? If not, why?

**Review by ArturGajowy — APPROVED** [review]

**Comment by ArturGajowy** [comment]
Please file an upstream PR, including the tempto config for running with simba drivers.
Also, this PR conflicts (alas, probably not on source level) with https://github.com/prestodb/presto/pull/6241. So care should be taken in whichever PR gets merged later.

**Comment by brian-rickman** [comment]
> > simba + kerberos will require new a new Docker image for Kerberos. It's going to take some time to get it to work. That will require new PRs for product-tests and docker images.
> 
> won't the cdh5-hive-kerberized image do? If not, why?

Simba says their drivers will only work with Presto if `http.server.authentication.krb5.service-name=HTTP` .  It's an issue with the HTTP library they use.

I realize this should be a trivial change, but I have made several attempts and have not yet gotten it to work on my dev box.

**Comment by ArturGajowy** [comment]
> Simba says their drivers will only work with Presto if http.server.authentication.krb5.service-name=HTTP . It's an issue with the HTTP library they use.

Since that goes into Presto's config.properties, we should be able to have that by adding a new profile, as each profile has its own presto config.properties, e.g. `presto/presto-product-tests/conf/presto/etc/singlenode-kerberized.properties`

**Comment by brian-rickman** [comment]
> > Simba says their drivers will only work with Presto if http.server.authentication.krb5.service-name=HTTP . It's an issue with the HTTP library they use.
> 
> Since that goes into Presto's config.properties, we should be able to have that by adding a new profile, as each profile has its own presto config.properties, e.g. presto/presto-product-tests/conf/presto/etc/singlenode-kerberized.properties

Except that the docker image has the kerberos principals and keytabs built into it, so using a different principal requires a new docker image.   Reference: https://github.td.teradata.com/center-for-hadoop/docker-images/blob/master/teradatalabs/cdh5-hive-kerberized/Dockerfile#L72

**Comment by ArturGajowy** [comment]
so the principal should be 'HTTP' instead of 'presto-server/presto-master@LABS.TERADATA.COM'?

Anyway, that should not be a big deal, since this additional layer would probably be < several MB. And we need another test profile anyway ;/

**Comment by brian-rickman** [comment]
> so the principal should be 'HTTP' instead of 'presto-server/presto-master@LABS.TERADATA.COM'?

I think that is the only change required, but I have not been able to get it to work yet.  I am still experimenting.

**Comment by brian-rickman** [comment]
Need to wait for other patches to be merged before I can cherry-pick this into sprint-41.

**Comment by brian-rickman** [comment]
Merged to sprint-41 branch.

**Pull request closed without merging**
